### ISSUE TITLE
[analyzer] Nullability of aliased types cannot be resolved in element model

### ISSUE THREAD

**Issue by dnys1** [dart-lang/sdk#54346]
Follow-up from conversation here: https://dart-review.googlesource.com/c/sdk/+/340643/comment/48461309_f14dfa96/

Currently, the analyzer provides `TypeAliasElement` for typedef'd types, but it does not provide the `nullabilitySuffix` applied to the _alias_, only the suffix applied to the aliased type.

For example, in the following code:

```dart
typedef A = String?;

final A a;
final A? b;
```

there is currently no way to distinguish fully between `a` and `b` in the element model.

cc @scheglov @bwilkerson

**Comment by TekExplorer** [comment]
I just ran into this while trying to create a writer.

I can get the underlying nullability suffix, so i can write the type directly, but its a bit silly that theres literally no way of knowing what the nullability suffix of the alias is unless i dig into the AST.

...which requires that I find the original usage, which really shouldnt be necessary,

It should be a relatively trivial fix.

```dart
abstract class InstantiatedTypeAliasElement {
  /// The alias element that is instantiated to produce a [DartType].
  TypeAliasElement get element;

  /// The type arguments with which the [element] was instantiated.
  /// This list will be empty if the [element] is not generic.
  List<DartType> get typeArguments;

  // ===== NEW ===== //
  /// doc comment here
  NullabilitySuffix get nullabilitySuffix;
  // ============== //
}

class InstantiatedTypeAliasElementImpl implements InstantiatedTypeAliasElement {
  @override
  final TypeAliasElementImpl element;

  @override
  final List<TypeImpl> typeArguments;

  // ===== NEW ===== //
  @override
  final NullabilitySuffix nullabilitySuffix;
  // ============== //

  InstantiatedTypeAliasElementImpl({
    required this.element,
    required this.typeArguments,
    // ===== NEW ===== //
    required this.nullabilitySuffix,
    // ============== //
  });
}

class TypeAliasElementImpl extends ElementImpl {
  @trackedIndirectly
  TypeImpl instantiateImpl({
    required List<TypeImpl> typeArguments,
    required NullabilitySuffix nullabilitySuffix,
  }) {
    // everywhere we instantiate `InstantiatedTypeAliasElementImpl`
    // add `nullabilitySuffix: nullabilitySuffix`
  }
}
```
`DartType.withNullability` would also need to be updated to support it.

**Comment by TekExplorer** [comment]
This also appears to affect `DartType.toString()` which ignores the alias entirely.

Maybe that's intended, but... i'm having trouble finding any tests for instantiated type aliases

In fact, `ElementDisplayStringBuilder._maybeWriteTypeAlias` looks to be using the _underlying_ `nullabilitySuffix` rather than the "actual" one.
_not to mention the example provided in the linked dart-review_

Now, admittedly, it should result in code thats just as functional... unless the user changes the typedef from `typedef Foo = Bar?;` to `typedef Foo = Bar;`

This kinda gets in the way of type aliases as intended. Plus resulting code are incorrectly documented (since its annotated with `?` when it shouldn't be, confusing the api.)

**Comment by TekExplorer** [comment]
Even @rrousselGit's `analyzer_buffer` package hits the same issue.

Like i said, it produced _valid_ code, but it's still semantically incorrect, especially if the underlying alias is changed in what could be a non-breaking fashion, (ex, generator-only types) except because of this you _need_ to regenerate your code (and might not realize!)

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/485583

**Comment by TekExplorer** [comment]
Nice!

Though, uh, it would be nice if we could have aliases for `Never`, `dynamic`, and `void`

That would probably require making them _not_ singletons.

Would that have to be a separate issue?

**Comment by scheglov** [comment]
Yeah, these I did not support from the beginning while adding type aliases support.
Yes, please, open a new issue, and we can discuss it there a little.
